File: data/src/main/java/org/cryptomator/data/cloud/crypto/CryptoCloudContentRepository.java
package org.cryptomator.data.cloud.crypto;

import android.content.Context;

import org.cryptomator.cryptolib.api.Cryptor;
import org.cryptomator.domain.CloudFolder;
import org.cryptomator.domain.exception.BackendException;
import org.cryptomator.domain.exception.CloudNodeAlreadyExistsException;
import org.cryptomator.domain.exception.FatalBackendException;
import org.cryptomator.domain.repository.CloudContentRepository;
import org.cryptomator.domain.usecases.ProgressAware;
import org.cryptomator.domain.usecases.cloud.DataSource;
import org.cryptomator.domain.usecases.cloud.DownloadState;
import org.cryptomator.domain.usecases.cloud.UploadState;
import org.cryptomator.util.Optional;
import org.cryptomator.util.Supplier;

import java.io.File;
import java.io.OutputStream;
import java.util.List;

import static java.lang.String.format;

class CryptoCloudContentRepository implements CloudContentRepository<CryptoCloud, CryptoNode, CryptoFolder, CryptoFile> {

	private final CryptoImplDecorator cryptoImpl;

	CryptoCloudContentRepository(Context context, CloudContentRepository cloudContentRepository, CryptoCloud cloud, Supplier<Cryptor> cryptor) {
		CloudFolder vaultLocation;
		try {
			vaultLocation = cloudContentRepository.resolve(cloud.getVault().getCloud(), cloud.getVault().getPath());
		} catch (BackendException e) {
			throw new FatalBackendException(e);
		}

		switch (cloud.getVault().getVersion()) {
			case 7:
				this.cryptoImpl = new CryptoImplVaultFormat7(context, cryptor, cloudContentRepository, vaultLocation, new DirIdCacheFormat7());
				break;
			case 6:
			case 5:
				this.cryptoImpl = new CryptoImplVaultFormatPre7(context, cryptor, cloudContentRepository, vaultLocation, new DirIdCacheFormatPre7());
				break;
			default:
				throw new IllegalStateException(format("No CryptoImpl for vault version %d.", cloud.getVault().getVersion()));
		}
	}

	@Override
	public synchronized CryptoFolder root(CryptoCloud cloud) throws BackendException {
		return cryptoImpl.root(cloud);
	}

	@Override
	public CryptoFolder resolve(CryptoCloud cloud, String path) throws BackendException {
		return cryptoImpl.resolve(cloud, path);
	}

	@Override
	public CryptoFile file(CryptoFolder parent, String name) throws BackendException {
		return cryptoImpl.file(parent, name);
	}

	@Override
	public CryptoFile file(CryptoFolder parent, String name, Optional<Long> size) throws BackendException {
		return cryptoImpl.file(parent, name, size);
	}

	@Override
	public CryptoFolder folder(CryptoFolder parent, String name) throws BackendException {
		return cryptoImpl.folder(parent, name);
	}

	@Override
	public boolean exists(CryptoNode node) throws BackendException {
		return cryptoImpl.exists(node);
	}

	@Override
	public List<CryptoNode> list(CryptoFolder folder) throws BackendException {
		return cryptoImpl.list(folder);
	}

	@Override
	public CryptoFolder create(CryptoFolder folder) throws BackendException {
		try {
			return cryptoImpl.create(folder);
		} catch (CloudNodeAlreadyExistsException e) {
			throw new CloudNodeAlreadyExistsException(folder.getName());
		}
	}

	@Override
	public CryptoFolder move(CryptoFolder source, CryptoFolder target) throws BackendException {
		try {
			return cryptoImpl.move(source, target);
		} catch (CloudNodeAlreadyExistsException e) {
			throw new CloudNodeAlreadyExistsException(target.getName());
		}
	}

	@Override
	public CryptoFile move(CryptoFile source, CryptoFile target) throws BackendException {
		try {
			return cryptoImpl.move(source, target);
		} catch (CloudNodeAlreadyExistsException e) {
			throw new CloudNodeAlreadyExistsException(target.getName());
		}
	}

	@Override
	public CryptoFile write(CryptoFile file, DataSource data, ProgressAware<UploadState> progressAware, boolean replace, long length) throws BackendException {
		return cryptoImpl.write(file, data, progressAware, replace, length);
	}

	@Override
	public void read(CryptoFile file, Optional<File> tmpEncryptedFile, OutputStream data, ProgressAware<DownloadState> progressAware) throws BackendException {
		cryptoImpl.read(file, data, progressAware);
	}

	@Override
	public void delete(CryptoNode node) throws BackendException {
		cryptoImpl.delete(node);
	}

	@Override
	public String checkAuthenticationAndRetrieveCurrentAccount(CryptoCloud cloud) throws BackendException {
		return cryptoImpl.currentAccount(cloud);
	}

	@Override
	public void logout(CryptoCloud cloud) throws BackendException {
		// empty
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/crypto/CryptoCloudContentRepository.kt
package org.cryptomator.data.cloud.crypto

import android.content.Context
import org.cryptomator.cryptolib.api.Cryptor
import org.cryptomator.domain.Cloud
import org.cryptomator.domain.CloudFile
import org.cryptomator.domain.CloudFolder
import org.cryptomator.domain.CloudNode
import org.cryptomator.domain.exception.BackendException
import org.cryptomator.domain.exception.CloudNodeAlreadyExistsException
import org.cryptomator.domain.exception.FatalBackendException
import org.cryptomator.domain.repository.CloudContentRepository
import org.cryptomator.domain.usecases.ProgressAware
import org.cryptomator.domain.usecases.cloud.DataSource
import org.cryptomator.domain.usecases.cloud.DownloadState
import org.cryptomator.domain.usecases.cloud.FileTransferState
import org.cryptomator.domain.usecases.cloud.UploadState
import java.io.File
import java.io.OutputStream
import java.util.function.Supplier

internal class CryptoCloudContentRepository(context: Context, cloudContentRepository: CloudContentRepository<Cloud, CloudNode, CloudFolder, CloudFile>, cloud: CryptoCloud, cryptor: Supplier<Cryptor>) :
	CloudContentRepository<CryptoCloud, CryptoNode, CryptoFolder, CryptoFile> {

	private var cryptoImpl: CryptoImplDecorator

	@Synchronized
	@Throws(BackendException::class)
	override fun root(cloud: CryptoCloud): CryptoFolder {
		return cryptoImpl.root(cloud)
	}

	override fun resolve(cloud: CryptoCloud, path: String): CryptoFolder {
		return cryptoImpl.resolve(cloud, path)
	}

	@Throws(BackendException::class)
	override fun file(parent: CryptoFolder, name: String): CryptoFile {
		return cryptoImpl.file(parent, name)
	}

	@Throws(BackendException::class)
	override fun file(parent: CryptoFolder, name: String, size: Long?): CryptoFile {
		return cryptoImpl.file(parent, name, size)
	}

	@Throws(BackendException::class)
	override fun folder(parent: CryptoFolder, name: String): CryptoFolder {
		return cryptoImpl.folder(parent, name)
	}

	@Throws(BackendException::class)
	override fun exists(node: CryptoNode): Boolean {
		return cryptoImpl.exists(node)
	}

	@Throws(BackendException::class)
	override fun list(folder: CryptoFolder): List<CryptoNode> {
		return cryptoImpl.list(folder)
	}

	@Throws(BackendException::class)
	override fun create(folder: CryptoFolder): CryptoFolder {
		return try {
			cryptoImpl.create(folder)
		} catch (e: CloudNodeAlreadyExistsException) {
			throw CloudNodeAlreadyExistsException(folder.name)
		}
	}

	@Throws(BackendException::class)
	override fun move(source: CryptoFolder, target: CryptoFolder): CryptoFolder {
		return try {
			cryptoImpl.move(source, target)
		} catch (e: CloudNodeAlreadyExistsException) {
			throw CloudNodeAlreadyExistsException(target.name)
		}
	}

	@Throws(BackendException::class)
	override fun move(source: CryptoFile, target: CryptoFile): CryptoFile {
		return try {
			cryptoImpl.move(source, target)
		} catch (e: CloudNodeAlreadyExistsException) {
			throw CloudNodeAlreadyExistsException(target.name)
		}
	}

	@Throws(BackendException::class)
	override fun write(file: CryptoFile, data: DataSource, progressAware: ProgressAware<UploadState>, replace: Boolean, size: Long): CryptoFile {
		return cryptoImpl.write(file, data, progressAware, replace, size)
	}

	@Throws(BackendException::class)
	override fun read(file: CryptoFile, encryptedTmpFile: File?, data: OutputStream, progressAware: ProgressAware<DownloadState>) {
		cryptoImpl.read(file, data, progressAware)
	}

	@Throws(BackendException::class)
	override fun associateThumbnails(list: List<CryptoNode>, progressAware: ProgressAware<FileTransferState>) {
		cryptoImpl.associateThumbnails(list, progressAware)
	}

	@Throws(BackendException::class)
	override fun delete(node: CryptoNode) {
		cryptoImpl.delete(node)
	}

	@Throws(BackendException::class)
	override fun checkAuthenticationAndRetrieveCurrentAccount(cloud: CryptoCloud): String {
		return cryptoImpl.currentAccount(cloud)
	}

	@Throws(BackendException::class)
	override fun logout(cloud: CryptoCloud) {
		// empty
	}

	init {
		val vaultLocation: CloudFolder = try {
			cloudContentRepository.resolve(cloud.vault.cloud, cloud.vault.path)
		} catch (e: BackendException) {
			throw FatalBackendException(e)
		}

		cryptoImpl = when (cloud.vault.format) {
			8 -> CryptoImplVaultFormat8(context, cryptor, cloudContentRepository, vaultLocation, DirIdCacheFormat7(), cloud.vault.shorteningThreshold)
			7 -> CryptoImplVaultFormat7(context, cryptor, cloudContentRepository, vaultLocation, DirIdCacheFormat7())
			6, 5 -> CryptoImplVaultFormatPre7(context, cryptor, cloudContentRepository, vaultLocation, DirIdCacheFormatPre7())
			else -> throw IllegalStateException(String.format("No CryptoImpl for vault format %d.", cloud.vault.format))
		}
	}
}
